### PR TITLE
feat: replaceable dashboard distro resource

### PR DESCRIPTION
**Pull request by shhdgit**
### What problem does this PR solve?

Support distribution build.

### What is changed and how it works?

New build process and inject distribution resource.

### Check List

Tests

- Manual test
build pd-server
image
run
image

Code changes

Side effects

Related changes

### Release note

```release-note
None
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -54,6 +54,10 @@ else
 	BUILD_CGO_ENABLED := 1
 endif
 
+ifneq ($(DASHBOARD_DISTRIBUTION_DIR),)
+	BUILD_TAGS += distro
```
Comment (breeswish) [inline Makefile:58]:
```suggestion
	BUILD_TAGS += dashboard-distro
```

File: Makefile [hunk 2]
```diff
@@ -125,6 +132,10 @@ dashboard-ui: export GO111MODULE=on
 dashboard-ui:
 	./scripts/embed-dashboard-ui.sh
 
+replace-distro-info:
```
Comment (breeswish) [inline Makefile:135]:
```suggestion
dashboard-replace-distro-info:
```

File: pkg/dashboard/distro/.gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+!/default_distro_info.go
```
Comment (breeswish) [inline pkg/dashboard/distro/.gitignore:1]: No need for this line?

File: pkg/dashboard/distro/distro.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+// limitations under the License.
+
+package distro
+
+import (
+	"github.com/pingcap/tidb-dashboard/pkg/utils/distro"
+)
+
+func init() {
+	distro.Replace(Resource)
```
Comment (breeswish) [inline pkg/dashboard/distro/distro.go:22]: Should we care about thread safety for this function?

File: scripts/embed-dashboard-ui.sh [hunk 5]
```diff
@@ -102,6 +108,16 @@ function compile_asset {
...
   echo '+ Build UI'
   cd "${BUILD_DIR}"
+
+  if [ -n "${DISTRIBUTION_DIR}" ]; then
+    echo '+ Generating distribution resource'
+    echo '  - Generating...'
+    DISTRIBUTION_DIR=${DISTRIBUTION_DIR} scripts/replace_distro_resource.sh
+    DISTRO_BUILD_TAG=1 scripts/generate_distro_info.sh
+
+    echo '+ Generating distribution resource complete'
```
Comment (breeswish) [inline scripts/embed-dashboard-ui.sh:118]:
```suggestion
    echo '+ Distribution resource generated'
```

File: pkg/dashboard/distro/.gitignore (hunk 3 above)
Comment (shhdgit) [reply pkg/dashboard/distro/.gitignore:1]: There's a `default*` rule in root/.gitignore.

File: pkg/dashboard/distro/distro.go (hunk 4 above)
Comment (shhdgit) [reply pkg/dashboard/distro/distro.go:22]: Yes, let me try.

**Review by baurine — COMMENTED** [review]
lgtm

File: pkg/dashboard/distro/default_distro_info.go [hunk 6]
```diff
@@ -0,0 +1,11 @@
+// +build !distro
+
+package distro
+
+// Resource declared the distro brand information
+var Resource = map[string]interface{}{
```
Comment (breeswish) [inline pkg/dashboard/distro/default_distro_info.go:6]: BTW why do we use `interface{}` instead of `string` here?

Comment (shhdgit) [reply pkg/dashboard/distro/default_distro_info.go:6]: This is a mistake, and I forget to fix it in the last update. One more round. Sorry for so many times review work.

**Comment by shhdgit** [comment]
PTAL, thanks!

**Review by breeswish — APPROVED** [review]

**Review by breeswish — APPROVED** [review]

**Comment by shhdgit** [comment]
/cc @HunDunDM 
PTAL, Thanks~

File: pkg/dashboard/distro/.gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (shhdgit) [reply pkg/dashboard/distro/.gitignore:1]: There's a default* rule in root/.gitignore.

**Review by HunDunDM — APPROVED** [review]

**Pull request merged**